stbt-docker

```diff
@@ -0,0 +1,184 @@
+        else:
+            raise
+
+    dockerfile = DOCKERFILE.format(base_image=base_image)
+    md5 = hashlib.md5()
+    md5.update(stbt_version)
+    md5.update(setup)
+    docker_image = "stb-tester-test-pack-%s" % md5.hexdigest()
+
+    try:
+        subprocess.check_output([DOCKER, 'inspect', docker_image],
+                                stderr=subprocess.STDOUT)
+        # Customised docker container is already cached
+        return docker_image
+    except subprocess.CalledProcessError:
+        pass
+    except OSError:
+        raise DockerNotInstalledError()
+
+    docker_build_context = {
+        'Dockerfile': dockerfile,
+        'setup': setup,
+    }
+    with dict_to_tarball(docker_build_context) as f:
+        subprocess.check_call(
+            [DOCKER, 'build', '-t', docker_image, '-'],
+            stdout=sys.stderr, stdin=f)
+
+    return docker_image
+
+
+def main(argv):
+    root = find_stbt_root()
+    if root is None:
+        sys.stderr.write(
+            "This command must be run within a test pack.  Couldn't find a "
+            ".stbt.conf in this or any parent directory.\n")
+        return 1
+
+    try:
+        docker_image = build_docker_image(root)
+    except DockerNotInstalledError:
+        sys.stderr.write(
+            "%s failed:\n%s requires docker.  Install docker, or if it is "
+            "already installed set $DOCKER to the location of the docker "
+            "executable.\n" % (" ".join(argv), os.path.basename(argv[0])))
+        return 1
+
+    if len(argv) > 1 and argv[1] == 'exec':
+        if argv[2:] == []:
+            args = ['python']
+        else:
+            args = argv[2:]
+    else:
+        args = ['stbt'] + argv[1:]
+
+    try:
+        os.makedirs(os.path.join(root, '.stbt/cache'))
+    except OSError:
+        # Directory already exists
+        pass
+
+    tty_arg = ['-t'] if sys.stdin.isatty() else []
+
+    innerdir = '/var/lib/stbt/test-pack/' + os.path.relpath(os.curdir, root)
+
+    # We use /bin/dash here because many programs (including make) get confused
+    # if they are running as pid 1.
+    os.execvp(
+        DOCKER,
+        [DOCKER, 'run', '-i'] + tty_arg + ['--rm', '-u', '0:0',
+         '-v', '%s:/var/lib/stbt/.cache' % os.path.join(root, '.stbt/cache'),
+         '-v', '%s:/var/lib/stbt/test-pack' % root,
+         '-e', 'STBT_CONFIG_FILE=/var/lib/stbt/test-pack/.stbt.conf',
+         '-e', 'PYTHONPATH=/usr/lib/stbt']
+        + shlex.split(os.environ.get('DOCKER_OPTS', '')) +
+        [docker_image, '/bin/dash', '-c',
+         'find /var/lib/stbt -xdev | grep -v test-pack | xargs chown -f %i:%i '
+         '&& sed -i s/stb-tester:x:1000:1000/stb-tester:x:%i:%i/ /etc/passwd '
+         '&& sed -i s/stb-tester:x:1000/stb-tester:x:%i/ /etc/group '
+         '&& cd %s && sudo -u stb-tester -E -- "$0" "$@"'
+         % (os.getuid(), os.getgid(), os.getuid(), os.getgid(), os.getgid(),
+            pipes.quote(innerdir))] + args)
```

Comment:
> `os.getuid` etc. won't be portable to Windows.

And it would be inappropriate for MacOS X with boot2docker.  The UIDs inside the VM will have no relation to on the MacOS host.

> We can figure out the portability later

Agreed.